Fix #15385: Added functionality to persist translator language on [CD - Translate Text].

ayushjaink8 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #15385.
2. This PR does the following: Adds functionality to persist translator language on Contributor dashboard page (Translate Text tab).

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

Part 1:

https://user-images.githubusercontent.com/75239620/167952817-d40742fb-2e3c-4976-8dfb-9f8294e7f2e8.mp4


Part 2:

https://user-images.githubusercontent.com/75239620/167952849-6e6a2d66-bca1-4d3f-bd2e-e941f27fd68c.mp4



## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).


## PR CI Checks:

1) ![image](https://user-images.githubusercontent.com/75239620/174726243-3ed48b19-61b8-4d88-a241-5b33adefa55f.png)

2) ![image](https://user-images.githubusercontent.com/75239620/175000731-95fbdf00-0b54-461f-b68f-2f9bcca82bfc.png)

3) ![image](https://user-images.githubusercontent.com/75239620/175322177-fcf06011-6a9d-473d-8ec3-49a8cfaacc58.png)


4) ![image](https://user-images.githubusercontent.com/75239620/175321928-c1e55f8f-14a7-44a4-a853-84c232614691.png)

5) ![image](https://user-images.githubusercontent.com/75239620/175461277-81448471-3c34-4115-9988-d61776c6230f.png)

ayushjaink8 committed:
added functionality to persist language on CD - Translate Text

ayushjaink8 committed:
removed redundant code

ayushjaink8 commented:
@sagangwee PTAL
PS: I have added `preferred_text_language_code` inside `UserSettingsModel` because we are already using this Model for `preferred_audio_language_code` and `preffered_site_language_code` i.e for exploration audio and site language preferences of the user. Later on, we can also add this setting in the `preferences` tab itself.

ayushjaink8 committed:
pip version error

ayushjaink8 committed:
Merge branch 'oppia:develop' into alang

sagangwee commented on core/storage/user/gae_models.py:
```diff
@@ -90,6 +90,9 @@ class UserSettingsModel(base_models.BaseModel):
         default=None, choices=[
             language['id'] for language in constants.SUPPORTED_AUDIO_LANGUAGES])
+    # Text language preference used for text translations.
+    preferred_text_language_code = datastore_services.StringProperty(
```
Maybe `preferred_translation_language_code` instead? To indicate in the name that this is the preferred language for submitted translations, whereas as the other language settings are for usage around the site.

ayushjaink8 replied:
Sounds good. I have updated this here and everywhere else.

sagangwee commented on core/controllers/contributor_dashboard.py:
```diff
@@ -524,3 +524,40 @@ def get(self):
+
+class TranslationPreferenceHandler(base.BaseHandler):
+    """Provides preffered translation language for the
+    Contribution dashboard page.
```
```suggestion
    """Provides the preferred translation language in the
    contributor dashboard.
```

ayushjaink8 replied:
Corrected.

sagangwee commented on core/controllers/contributor_dashboard.py:
```diff
@@ -524,3 +524,40 @@ def get(self):
+                user_settings.preferred_text_language_code)
+        })
+        self.render_json(self.values['preferred_text_language_code'])
+        # self.render_json('ak')
```
Remove.

ayushjaink8 replied:
Removed.

sagangwee commented on core/controllers/contributor_dashboard.py:
```diff
@@ -524,3 +524,40 @@ def get(self):
+    def get(self):
+        """Handles GET requests."""
+        user_settings = user_services.get_user_settings(self.user_id)
+        self.values.update({
+            'preferred_text_language_code': (
+                user_settings.preferred_text_language_code)
+        })
+        self.render_json(self.values['preferred_text_language_code'])
```
We typically return the dictionary of values. See above for examples.

ayushjaink8 replied:
Updated this.

sagangwee commented on main.py:
```diff
@@ -282,6 +282,9 @@ def get_redirect_route(
         r'%s/<opportunity_type>' % feconf.CONTRIBUTOR_OPPORTUNITIES_DATA_URL,
         contributor_dashboard.ContributionOpportunitiesHandler),
+    get_redirect_route(
+        r'/prefferedtranslationlanguage',
```
Typo.

ayushjaink8 replied:
Corrected (here and elsewhere).

sagangwee commented on core/templates/pages/contributor-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -215,6 +215,29 @@ export class ContributionOpportunitiesBackendApiService {
   }
+
+  async saveUserLanguagePreferenceAsync(language_code: string):
+  Promise<unknown> {
```
Should this return a void Promise? 

```suggestion
  Promise<void> {
```

ayushjaink8 replied:
I have changed this `<Object>` instead. `<void>` was throwing an error.
![image](https://user-images.githubusercontent.com/75239620/168761541-c94b47c9-9cc3-474b-8c7f-f62cebb2c0a1.png)

sagangwee replied:
I think you need to specify the Promise return type. Added a comment.

sagangwee commented on core/templates/pages/contributor-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -215,6 +215,29 @@ export class ContributionOpportunitiesBackendApiService {
+
+  async fetchUserLanguagePreferenceAsync():
+  Promise<string|null> {
+    try {
+      const response = await this.http.get<string>(
+        '/prefferedtranslationlanguage').toPromise();
+      return response;
+    } catch {
+      return null;
```
What exception is being caught here?

ayushjaink8 replied:
I wrote this in a try-catch block to deal with errors during this promise. I have now changed the way a little bit. Kindly, have another look.

ayushjaink8 replied:
I have changed this back to what it was because our code might break if we throw an error over here. Actually, I am following the general pattern over here which is being followed elsewhere in this file.
E.g. lines 191-218 https://github.com/oppia/oppia/blob/02db474efed7bfb91864c04ba3db2c82969fb5aa/core/templates/pages/contributor-dashboard-page/services/contribution-opportunities-backend-api.service.ts#L191-L218

sagangwee commented on core/domain/user_services.py:
```diff
@@ -1236,6 +1238,21 @@ def update_preferred_audio_language_code(
+    """
+    user_settings = get_user_settings(user_id, strict=True)
+    user_settings.preferred_text_language_code = (
+        preferred_text_language_code)
```
Should we validate the supplied language code?

ayushjaink8 replied:
I have added a validator in the `POST` argument of `TranslationPreferenceHandler`.

sagangwee commented on core/templates/pages/contributor-dashboard-page/translation-language-selector/translation-language-selector.component.ts:
```diff
@@ -89,17 +90,39 @@ export class TranslationLanguageSelectorComponent implements OnInit {
 
+  async saveCurrentSelectedLanguage(
+    languageCode: string): Promise<unknown> {
+    console.log(languageCode);
```
Remove.

sagangwee commented on core/templates/pages/contributor-dashboard-page/translation-language-selector/translation-language-selector.component.ts:
```diff
@@ -29,6 +29,7 @@ import { FeaturedTranslationLanguage } from
   'domain/opportunity/featured-translation-language.model';
 import { LanguageUtilService } from 'domain/utilities/language-util.service';
 import { TranslationLanguageService } from 'pages/exploration-editor-page/translation-tab/services/translation-language.service';
+import { i } from 'mathjs';
```
Where is this used?

ayushjaink8 replied:
Added by mistake. I have removed this.

sagangwee commented on core/templates/pages/contributor-dashboard-page/translation-language-selector/translation-language-selector.component.ts:
```diff
@@ -89,17 +90,39 @@ export class TranslationLanguageSelectorComponent implements OnInit {
+      .saveUserLanguagePreferenceAsync(languageCode);
+  }
+
+  populateSavedLanguageSelection(languageCode: string): void {
+    this.setActiveLanguageCode.emit(languageCode);
+    this.languageSelection = this.languageIdToDescription[languageCode];
+  }
```
This method isn't particular to the saved language, so we can probably rename to just `populateLanguageSelection` and reuse it in `selectOption`.

ayushjaink8 replied:
Done.

sagangwee reviewed (changes requested):
@ayushjaink8 Thanks, took a pass!

ayushjaink8 committed:
requested changes

ayushjaink8 committed:
Merge branch 'alang' of https://github.com/ayushjaink8/oppia into alang

ayushjaink8 committed:
Merge branch 'oppia:develop' into alang

ayushjaink8 commented:
@sagangwee I have made the requested changes and replied to all of your comments. Kindly, have another look.

ayushjaink8 committed:
fixed few errors

sagangwee commented on core/controllers/contributor_dashboard.py:
```diff
@@ -524,3 +524,42 @@ def get(self):
+
+
+class TranslationPreferenceHandler(base.BaseHandler):
+    """Provides the prefered translation language in the
```
Typo in prefered*

ayushjaink8 replied:
oh! My bad. I have fixed this.

sagangwee commented on core/controllers/contributor_dashboard.py:
```diff
@@ -524,3 +524,42 @@ def get(self):
+    def get(self):
+        """Handles GET requests."""
+        user_settings = user_services.get_user_settings(self.user_id)
+        self.values.update({
+            'preferred_translation_language_code': (
+                user_settings.preferred_translation_language_code)
+        })
+        self.render_json(self.values)
```
I don't think it's necessary to update the values—you can return a new dictionary directly:

```suggestion
        self.render_json({
          'preferred_translation_language_code': 
                user_settings.preferred_translation_language_code
        })
```

ayushjaink8 replied:
I have updated this.

sagangwee commented on core/storage/user/gae_models.py:
```diff
@@ -90,6 +90,9 @@ class UserSettingsModel(base_models.BaseModel):
     preferred_audio_language_code = datastore_services.StringProperty(
         default=None, choices=[
             language['id'] for language in constants.SUPPORTED_AUDIO_LANGUAGES])
+    # Text language preference used for text translations.
```
Let's make it clearer that this is the preference when submitting translations, e.g.

```suggestion
    # Language preference when submitting text translations in the contributor dashboard.
```

ayushjaink8 replied:
I have updated this.

sagangwee commented on core/templates/pages/contributor-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -215,6 +219,28 @@ export class ContributionOpportunitiesBackendApiService {
+    };
+    const body = new FormData();
+    body.append('payload', JSON.stringify(postData));
+    const response = this.http.post(
+      '/preferredtranslationlanguage', body).toPromise();
+    return response;
```
The post call doesn't seem to return anything, so I think we can return void by specifying `<void>`:

```suggestion
    return this.http.post<void>(
      '/preferredtranslationlanguage', body).toPromise();
```

ayushjaink8 replied:
Actually, I am facing issues with this. We need to return something in the post call. Otherwise, we will get the following error during `self.post_json()`:
![image](https://user-images.githubusercontent.com/75239620/169136137-9ef9d2a3-21ff-409c-bb3e-9dffe924e48a.png)


To avoid this, I have added `self.render_json({})` in the post call. We are doing this at other locations as well.
For example:
https://github.com/oppia/oppia/blob/0af67f015ab36e2cb2fc5efbeda157a4528a6a1d/core/controllers/profile.py#L537-L560

Now, what we should do over here? Is there any other alternative way to deal with this issue?

sagangwee commented on core/templates/pages/contributor-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -215,6 +219,28 @@ export class ContributionOpportunitiesBackendApiService {
   }
+
+  async saveUserLanguagePreferenceAsync(languageCode: string):
+  Promise<Object> {
```
I think we can return a void Promise.

ayushjaink8 replied:
I think we cannot. Kindly refer to my previous comment.
Comment link: https://github.com/oppia/oppia/pull/15460#discussion_r876249148

sagangwee replied:
I'm not sure I understand the error you were facing in the comment. After returning `self.render_json({})` in the backend, what error do you run into if you try to return `Promise<void>` here? e.g. we return `self.render_json({})` in the backend [here](https://github.com/oppia/oppia/blob/develop/core/controllers/creator_dashboard.py#L283), and return a void Promise in the corresponding backend service [here](https://github.com/oppia/oppia/blob/develop/core/templates/domain/creator_dashboard/creator-dashboard-backend-api.service.ts#L191).

ayushjaink8 replied:
Thank you for the reference. It was a silly mistake on my end. I have replaced this with `<void>`.

sagangwee commented on core/templates/pages/contributor-dashboard-page/services/contribution-opportunities.service.ts:
```diff
@@ -127,6 +137,15 @@ export class ContributionOpportunitiesService {
     return this._getTranslatableTopicNamesAsync();
   }
 
+  async getUserLanguagePreferenceAsync(): Promise<string> {
+    return this._getUserLanguagePreferenceAsync();
+  }
+
+  async saveUserLanguagePreferenceAsync(
+      languageCode: string): Promise<Object> {
+    return this._saveUserLanguagePreferenceAsync(languageCode);
+  }
```
Why not keep these function names consistent with the property name, e.g. `getPreferredTranslationLanguage`?

ayushjaink8 replied:
I have updated this.

sagangwee commented on core/templates/pages/contributor-dashboard-page/services/contribution-opportunities.service.ts:
```diff
@@ -90,6 +90,16 @@ export class ContributionOpportunitiesService {
       .fetchTranslatableTopicNamesAsync();
   }
 
+  private async _getUserLanguagePreferenceAsync() {
+    return this.contributionOpportunitiesBackendApiService
+      .fetchUserLanguagePreferenceAsync();
+  }
+
+  private async _saveUserLanguagePreferenceAsync(languageCode: string) {
+    return this.contributionOpportunitiesBackendApiService
+      .saveUserLanguagePreferenceAsync(languageCode);
+  }
```
These methods don't seem necessary.

ayushjaink8 replied:
I have removed these methods.

sagangwee commented on core/templates/pages/contributor-dashboard-page/translation-language-selector/translation-language-selector.component.ts:
```diff
@@ -89,17 +89,36 @@ export class TranslationLanguageSelectorComponent implements OnInit {
     this.dropdownShown = !this.dropdownShown;
   }
 
+  async saveCurrentSelectedLanguage(
+      languageCode: string): Promise<Object> {
+    return await this.contributionOpportunitiesBackendApiService
+      .saveUserLanguagePreferenceAsync(languageCode);
+  }
```
This method doesn't seem necessary? We can call the API directly in `selectOption`.

ayushjaink8 replied:
I have updated this.

sagangwee reviewed (changes requested):
@ayushjaink8 Took another pass. Note, you'll also need to add new tests.

ayushjaink8 committed:
implemented requested changes and fixed backend tests

ayushjaink8 commented:
@sagangwee I have addressed all your suggestions/changes. I have also fixed and added backend tests. Kindly, have a look. Meanwhile, I will work on front-end tests.

ayushjaink8 committed:
fixed lint errors and python type check errors

ayushjaink8 committed:
fixed backend tests

ayushjaink8 committed:
fixed takeout_service_test.py error

ayushjaink8 committed:
fixed front-end tests and added new tests

ayushjaink8 committed:
Merge branch 'oppia:develop' into alang

ayushjaink8 committed:
fixed lint errors

ayushjaink8 committed:
fixed e2e coverage and typescipt tests failing.

ayushjaink8 commented:
@sagangwee PTAL

ayushjaink8 committed:
Merge branch 'oppia:develop' into alang

sagangwee commented on core/templates/pages/contributor-dashboard-page/services/contribution-opportunities-backend-api.service.spec.ts:
```diff
@@ -371,4 +371,107 @@ describe('Contribution Opportunities backend API service', function() {
+    })
+  );
+
+  it('should return empty string if \'preferredtranslationlanguage call fails',
```
```suggestion
  it('should return empty string if \'preferredtranslationlanguage\' call fails',
```

ayushjaink8 replied:
Fixed.

sagangwee commented on core/templates/pages/contributor-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -215,6 +219,36 @@ export class ContributionOpportunitiesBackendApiService {
+      return response.preferred_translation_language_code;
+    } catch {
+      return '';
+    }
```
Back to https://github.com/oppia/oppia/pull/15460#discussion_r873907530, it doesn't look like there is an exception that is being caught here. The other examples in this file seem to have added the try blocks to catch specific exceptions, e.g. `response.topic_names.unshift('All');` will raise an error if the `topic_names` array is empty, but in this case, it's probably unnecessary since we're just returning the response value directly.

ayushjaink8 replied:
Ohk! I got your point. Actually, I was adding these try-catch block in order to deal with server-side errors like what if our backend server is down and there is a 404 error at `'/preferredtranslationlanguage'` or when this get request fails due to any other reason. I have now removed this try-catch block and added a catch on the promise object itself. Does this look good to you?

sagangwee commented on core/templates/pages/contributor-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -215,6 +219,36 @@ export class ContributionOpportunitiesBackendApiService {
+    return this.http.post<Object>(
+      '/preferredtranslationlanguage', body).toPromise().then(
+      data => {
+        return data;
```
Since nothing usable is actually being returned, we can omit the `then` clause and just call `catch` on the Promise object below to parse the error.

ayushjaink8 replied:
Edited.

sagangwee commented on core/templates/pages/contributor-dashboard-page/translation-language-selector/translation-language-selector.component.spec.ts:
```diff
@@ -183,6 +185,44 @@ describe('Translation language selector', () => {
+  it('should display the preferred language when the preferred' +
+    ' language is defined', async(() => {
+    preferredLanguageCode = 'en';
+    let languageDescription = component.languageIdToDescription.en;
```
Prefer `const` if the variable is not being reassigned. Also, let's use the description from `constants.SUPPORTED_AUDIO_LANGUAGES.en` directly as the source of truth.

ayushjaink8 replied:
`constants.SUPPORTED_AUDIO_LANGUAGES.en` is not defined since we are not having key-value pairs in `SUPPORTED_AUDIO_LANGUAGES` so I have used `find()` method in order to deal with this.

Edit: I have also used a `@ts-ignore` comment to avoid the `TS2532` error: Object is possibly 'undefined'.

sagangwee commented on core/templates/pages/contributor-dashboard-page/translation-language-selector/translation-language-selector.component.spec.ts:
```diff
@@ -183,6 +185,44 @@ describe('Translation language selector', () => {
+    spyOn(component.setActiveLanguageCode, 'emit').and.callFake(
+      (languageCode) => {
+        component.activeLanguageCode = (
+          languageCode ? languageCode : null);
```
Why is this ternary necessary?

ayushjaink8 replied:
Actually, I was facing problems in resolving the front-end tests so I was trying different things in order to debug the issue. I thought it would solve the issue and it did work but now I have realized that the main issue was in the `contributionOpportunitiesBackendApiServiceStub` where I have added an `if` statement (in the `getPreferredTranslationLanguageAsync` method to match with the actual function present inside `ngOnInit()`).
This ternary is irrelevant so I have removed this now.

ayushjaink8 replied:
I added this ternary to remove the Typescript error `TS2322: Type 'string | undefined' is not assignable to type 'string | null'`. I hope it makes sense.

sagangwee replied:
You might need to declare the `languageCode` param as a string type, e.g.`callFake((languageCode: string)` to satisfy the compiler.

sagangwee commented on core/templates/pages/contributor-dashboard-page/translation-language-selector/translation-language-selector.component.spec.ts:
```diff
@@ -195,4 +235,25 @@ describe('Translation language selector', () => {
+
+  it('should indiacate selection and save the language' +
+    ' on selecting a new language', () => {
+    let selectedLanguage = 'fr';
```
Prefer `const`.

ayushjaink8 replied:
Edited.

sagangwee reviewed (changes requested):
@ayushjaink8 Thanks, took a pass.

ayushjaink8 committed:
Merge branch 'oppia:develop' into alang

ayushjaink8 committed:
fixed known bugs and applied requested changes

ayushjaink8 committed:
removed irrelevant comments

ayushjaink8 committed:
fixed lint and typescript errors

ayushjaink8 committed:
fixed e2e tests errors and added a new e2e test

ayushjaink8 committed:
fixed lint errors

ayushjaink8 commented:
@sagangwee I have replied to all of your comments. Fixed e2e errors and added one e2e test as well. Kindly have a look.

PS: I have made some changes in the `TranslationPreferenceHandler` in the `contributor_dashboard.py` to resolve the e2e errors.

ayushjaink8 committed:
fixed typescript check error

sagangwee commented on core/controllers/contributor_dashboard.py:
```diff
@@ -524,3 +524,46 @@ def get(self):
+    @acl_decorators.open_access
+    def get(self):
+        """Handles GET requests."""
+        if self.user_id:
+            user_settings = user_services.get_user_settings(self.user_id)
+            self.render_json({
+                'preferred_translation_language_code': (
+                    user_settings.preferred_translation_language_code)
+            })
+        else:
+            self.render_json({'preferred_translation_language_code': ''})
```
```suggestion
          user_settings = user_services.get_user_settings(self.user_id)
          if user_settings is None:
            return self.render_json({'preferred_translation_language_code': ''})
          return self.render_json({
              'preferred_translation_language_code': (
                  user_settings.preferred_translation_language_code)
          })
```

sagangwee commented on core/templates/pages/contributor-dashboard-page/translation-language-selector/translation-language-selector.component.spec.ts:
```diff
@@ -183,6 +188,47 @@ describe('Translation language selector', () => {
+    ' language is defined', async(() => {
+    component.activeLanguageCode = null;
+    preferredLanguageCode = 'en';
+    // This throws "TS2532". We need to suppress this error because
+    // "constants.SUPPORTED_AUDIO_LANGUAGES" is always defined. Also, it must
+    // contain description for language code 'en' and if it doesn't
+    // then this test should fail.
+    // @ts-ignore
+    const languageDescription = constants.SUPPORTED_AUDIO_LANGUAGES.find(
+      e => e.id === 'en').description;
```
Can we use `LanguageUtilService.getAudioLanguageDescription()` here?

ayushjaink8 replied:
@sagangwee Yes we can but it will be implemented like this:
```
const languageDescription = new LanguageUtilService(new BrowserCheckerService(
      new WindowRef)).getAudioLanguageDescription(preferredLanguageCode);
```
For this, we need to import `LanguageUtilService`, `BrowserCheckerSerivice`, and `WindowRef`.

Should I update this?

sagangwee replied:
I see. What you have then is probably fine, then. Thanks.

sagangwee commented on core/tests/protractor_desktop/contributorDashboard.js:
```diff
@@ -114,13 +114,34 @@ describe('Contributor dashboard page', function() {
   });
 
   it('should allow user to switch to translate text tab', async function() {
+    await users.login(USER_EMAILS[0]);
     await contributorDashboardPage.get();
     await contributorDashboardPage.navigateToTranslateTextTab();
     await contributorDashboardTranslateTextTab.changeLanguage(GERMAN_LANGUAGE);
     await contributorDashboardTranslateTextTab.expectSelectedLanguageToBe(
       GERMAN_LANGUAGE);
+    await users.logout();
```
Why were these changes necessary here?

ayushjaink8 replied:
Because if we try to change the language without login, we will encounter a console log error and this test would fail. We make a post request whenever a user changes a language from the dropdown (which will return an error if the user is not logged in).

seanlip replied:
But this seems odd. We should make sure that the behaviour works correctly for non-logged-in users, too. If it wasn't throwing an error before, it should not throw one now.

ayushjaink8 replied:
I have removed this now. I have added a check in the front-end itself to make a post request for saving the language preference only when user is logged-in.

sagangwee reviewed (approved):
@ayushjaink8 LGTM, minus a few comments. Thanks for making this change!

vojtechjelinek commented on core/controllers/contributor_dashboard.py:
```diff
@@ -524,3 +524,46 @@ def get(self):
+                'schema': {
+                    'type': 'basestring',
+                    'validators': [{
+                        'id': 'is_supported_audio_language_code'
```
This is a bit confusing, should we rename the function to smth like `is_supported_language_code`?

ayushjaink8 replied:
As of now, we are using `SUPPORTED_AUDIO_LANGUAGES` for both written translations and voice-over translations and also for setting the audio language for the exploration. `is_supported_language_code` uses `SUPPORTED_AUDIO_LANGUAGES` to validate. If we rename this validator then I believe we need to make changes to a lot of places. I think this will be out of scope for this PR, we can create a new issue for the renaming of all these validators and make sure that the name clearly defines the purpose.

vojtechjelinek replied:
Yes, please create a new issue.

ayushjaink8 replied:
I have created a new issue.
Issue Link: https://github.com/oppia/oppia/issues/15559

vojtechjelinek replied:
Add a TODO comment then.

ayushjaink8 replied:
I have added a TODO comment just above the "HANDLER_ARGS_SCHEMAS".

vojtechjelinek commented on core/controllers/contributor_dashboard.py:
```diff
@@ -524,3 +524,46 @@ def get(self):
+        }
+    }
+
+    @acl_decorators.open_access
```
Why is this open access? Shouldn't this be limited to logged-in users?

ayushjaink8 replied:
Yes, it should be limited to the logged-in users but the "contributor dashboard page" can be accessed by anyone and anyone can select a language to see the opportunities, and whenever one selects, a post request will be sent to save the preference. Now, I am not sure if we should use `@acl_decorators.can_manage_own_account` or `@acl_decorators.open_access`.
Also, If I use `@acl_decorators.can_manage_own_account` then it may lead to errors because when guest users try to change the language, the post request will throw an error. I am facing this particular error in the e2e tests as well.

ayushjaink8 replied:
@sagangwee @vojtechjelinek  Is there any way to distinguish between a logged-in user and a guest user in the front-end itself? 
I will then add an if-else statement there itself so that the post request will not be sent if the user is not logged-in to avoid the error.

seanlip replied:
I don't know offhand, but if you look at the logged-in and logged-out views, the CD is showing different things, right?

So maybe look into how it's doing that, and you might be able to use that info as inspiration.

It does seem to me that this handler should assume the user is logged in (and the backend tests should be updated accordingly).

ayushjaink8 replied:
Thanks for this tip. This was really helpful.  
I have now changed this to `@acl_decorators.can_manage_own_account` and added an if-else statement in the front-end to make the post call for saving user preference only when the user is logged in by using `UserService`.

seanlip replied:
It still looks like open_access in my Files Changed view.

ayushjaink8 replied:
I think now, these changes would be visible to you. I have made another commit for the same.

(PS: I don't know why this commit is not updated on Github. It ain't reflecting at my end too so I made another commit.)

vojtechjelinek commented on core/domain/user_services.py:
```diff
@@ -1236,6 +1238,22 @@ def update_preferred_audio_language_code(
     _save_user_settings(user_settings)
 
 
+def update_preferred_translation_language_code(
+        user_id, preferred_translation_language_code):
```
```suggestion
def update_preferred_translation_language_code(
    user_id, preferred_translation_language_code
):
```

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -215,6 +219,32 @@ export class ContributionOpportunitiesBackendApiService {
     }
   }
+
+  async savePreferredTranslationLanguageAsync(languageCode: string):
+  Promise<void> {
```
```suggestion
  async savePreferredTranslationLanguageAsync(
      languageCode: string
  ): Promise<void> {
```

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -215,6 +219,32 @@ export class ContributionOpportunitiesBackendApiService {
+      });
+  }
+
+  async getPreferredTranslationLanguageAsync():
+  Promise<string> {
```
```suggestion
  async getPreferredTranslationLanguageAsync(
  ): Promise<string> {
```

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/translation-language-selector/translation-language-selector.component.spec.ts:
```diff
@@ -27,6 +27,7 @@ import { ContributionOpportunitiesBackendApiService } from
 import { FeaturedTranslationLanguage } from 'domain/opportunity/featured-translation-language.model';
 import { TranslationLanguageService } from 'pages/exploration-editor-page/translation-tab/services/translation-language.service';
 import { EventEmitter } from '@angular/core';
+import constants from 'assets/constants';
```
Use AppConstants.

ayushjaink8 replied:
Updated.

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

ayushjaink8 committed:
Update core/domain/user_services.py

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

ayushjaink8 committed:
Update core/templates/pages/contributor-dashboard-page/services/contribution-opportunities-backend-api.service.ts

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

ayushjaink8 committed:
Update core/templates/pages/contributor-dashboard-page/services/contribution-opportunities-backend-api.service.ts

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

ayushjaink8 committed:
Update core/controllers/contributor_dashboard.py

Co-authored-by: Sagang Wee <9004726+sagangwee@users.noreply.github.com>

ayushjaink8 committed:
accomodated suggested changes

ayushjaink8 committed:
Merge branch 'oppia:develop' into alang

ayushjaink8 commented:
@vojtechjelinek I have replied to all of your comments and incorporated your suggestions. Kindly take a look.

seanlip commented on core/domain/user_domain.py:
```diff
@@ -95,6 +96,8 @@ class UserSettings:
             preferences specified by the user.
         preferred_site_language_code: str or None. System language preference.
         preferred_audio_language_code: str or None. Audio language preference.
+        preferred_translation_language_code: str or None. Translation
```
Explain more what this is, specifically that it pertains to the contributor dashboard. Otherwise it sounds like the language the user wants the site translated into.

ayushjaink8 replied:
I have updated this (elsewhere too) and added that this pertains to the contributor dashboard.

seanlip replied:
I think you need to be more specific, "pertains to the contributor dashboard" still isn't clear. Think about a new developer looking at this code, who has limited context. Can you explain clearly what preferred_translation_language_code represents and what it's used for so that there is no possibility of confusion?

(I wasn't expecting you to quote my words verbatim -- think about what's best for the end user (the developer in this case).)

ayushjaink8 replied:
I have edited this description to
```
Text Translation language preference of the translator that persists on the contributor dashboard.
```
Does this sound good?

(I included that it is a user preference who wants to translate text and it persists on the contributor dashboard". Persistence on the contributor dashboard is the primary use of this attribute.)

seanlip commented on core/templates/pages/contributor-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -215,6 +219,33 @@ export class ContributionOpportunitiesBackendApiService {
+    const postData: {'language_code': string} = {
+      language_code: languageCode,
+    };
+    const body = new FormData();
```
Do you actually need to create a FormData() here? I looked through the codebase and you can just pass the postData in a dict, I think? (see admin-backend-api.service.ts, playthrough-backend-api.service.ts)

ayushjaink8 replied:
I have updated this and now I am just passing the postData in a dict directly.

seanlip commented on core/templates/pages/contributor-dashboard-page/translation-language-selector/translation-language-selector.component.ts:
```diff
@@ -89,17 +89,32 @@ export class TranslationLanguageSelectorComponent implements OnInit {
       'Select a language...'
     );
+
+    this.contributionOpportunitiesBackendApiService
```
I'm a bit confused about why you're doing this in a separate call to the backend.

Couldn't it be combined with the initial call that's used to fetch the data for the CD? Otherwise it increases the overall loading/stabilization time for the page.

ayushjaink8 replied:
I am not sure how we can do that? I mean contributor dashboard does not directly load to the "Translate Text" tab so we do not call for any data related to the "Translate Text" tab initially and only load data for "My Contributions" tab. When the user switches the tab then only we render this translate-language-selector component and call for the preferred translation language.
Should I make this call from somewhere else? If yes, then from where and how I will be able to get the data over here in this component ( translate-language-selector ) because at the end I need to change the value of this component so that it should further call for the opportunities available in that language.

seanlip replied:
Hm, I'm looking again at the overall code for the contributor dashboard, and I actually see this pattern quite a bit -- lots of small calls for little pieces of data. So I suppose this follows the existing trend and is probably OK, given the context.

@sagangwee -- just wondering though, have you considered this point and do you think this might be impacting page speed?

sagangwee replied:
That's an interesting point. I suppose we could batch together any CD related data on initial page component load into one API endpoint, e.g. combine this fetch with the `ContributionOpportunitiesService.getTranslatableTopicNamesAsync()` [here](https://github.com/oppia/oppia/blob/develop/core/templates/pages/contributor-dashboard-page/contributor-dashboard-page.component.ts#L216) and then fanout the data to the child components.

Though this small optimization might be dwarfed by all the other network requests we're making, e.g. we make at least 3 calls to `UserService` in the [contributor dashboard page component](https://github.com/oppia/oppia/blob/develop/core/templates/pages/contributor-dashboard-page/contributor-dashboard-page.component.ts#L162).

It also looks like we're repeating a bunch of network requests on initial page load, e.g. multiple calls to `/gettranslatabletopicnames`:
![Screen Shot 2022-06-14 at 6 05 22 PM](https://user-images.githubusercontent.com/9004726/173698031-7f67afb3-9ede-45d5-b625-c2a545093892.png)

So there might be some Angular improvements we can make there. Will track in https://github.com/oppia/oppia/issues/15588.

Assessing the performance impact probably needs a bit of work, so I'm inclined to leave the API usage as is until we can get better impact measurements.

seanlip reviewed (changes requested):
Thanks @ayushjaink8! I can see this being very useful for our translator community, so thank you for working on it!

Took a pass and left some comments, PTAL.

kevintab95 commented on core/templates/pages/contributor-dashboard-page/translation-language-selector/translation-language-selector.component.spec.ts:
```diff
@@ -183,6 +188,46 @@ describe('Translation language selector', () => {
+      fixture.detectChanges();
+      expect(component.setActiveLanguageCode.emit)
+        .toHaveBeenCalledWith(preferredLanguageCode);
+      expect(component.activeLanguageCode).toBe(preferredLanguageCode);
+      expect(component.languageSelection).toBe(languageDescription);
```
Should we add a pre-check for this before line 208?

ayushjaink8 replied:
I have specifically declared `activeLanguageCode` and `languageSelection` in this test.
```
    component.activeLanguageCode = null;
    component.languageSelection = '';
``` 
So now, I think having a pre-check won't be required.

kevintab95 commented on core/templates/pages/contributor-dashboard-page/translation-language-selector/translation-language-selector.component.spec.ts:
```diff
@@ -183,6 +188,46 @@ describe('Translation language selector', () => {
+    component.ngOnInit();
+
+    fixture.detectChanges();
+    expect(component.languageSelection).toBe('Select a language...');
```
Should we add a pre-check for this before line 224?

ayushjaink8 replied:
I have specifically declared `activeLanguageCode` and `languageSelection` in this test.
```
    component.activeLanguageCode = null;
    component.languageSelection = '';
``` 
So now, I think having a pre-check won't be required.

kevintab95 commented on core/templates/pages/contributor-dashboard-page/translation-language-selector/translation-language-selector.component.spec.ts:
```diff
@@ -195,4 +240,25 @@ describe('Translation language selector', () => {
       expect(component.languageSelection).toBe('français (French)');
     });
+
+  it('should indiacate selection and save the language' +
```
typo: indiacate -> indicate

ayushjaink8 replied:
Corrected. Thanks!

kevintab95 reviewed (changes requested):
Thanks @ayushjaink8! Left couple of minor comments on 2 codeowner files. PTAL

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/translation-language-selector/translation-language-selector.component.spec.ts:
```diff
@@ -183,6 +188,46 @@ describe('Translation language selector', () => {
+    ' language is defined', async(() => {
+    component.activeLanguageCode = null;
+    preferredLanguageCode = 'en';
+    // This throws "TS2532". We need to suppress this error because
+    // "constants.SUPPORTED_AUDIO_LANGUAGES" is always defined. Also, it must
+    // contain description for language code 'en' and if it doesn't
+    // then this test should fail.
+    // @ts-ignore
```
Is this ignore still needed?

ayushjaink8 replied:
Actually, I am getting a "TS2523" error if I try to find language description from AppConstants like this: 
```
const languageDescription = AppConstants.SUPPORTED_AUDIO_LANGUAGES.find(
      e => e.id === 'en').description;
```
But now I have removed this ts-ignore comment by using optional chaining like this:
```
const languageDescription = AppConstants.SUPPORTED_AUDIO_LANGUAGES.find(
      e => e.id === 'en')?.description ?? '';
```

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

SAEb-ai reviewed (changes requested):
@ayushjaink8 Blocking because of [@seanlip's comment](https://github.com/oppia/oppia/pull/15460/files?file-filters%5B%5D=.js&file-filters%5B%5D=.py&file-filters%5B%5D=.ts&owned-by%5B%5D=SAEb-ai&show-viewed-files=true#r893031087). Feel free to reassign me once this is resolved. Thanks!!

ayushjaink8 committed:
Merge branch 'develop' into alang

ayushjaink8 committed:
fixed e2e error for logged-in and non-logged-in users by adding UserService

ayushjaink8 committed:
removed typescript errors

ayushjaink8 committed:
removed typescript errors

ayushjaink8 committed:
fixed typescript errors

ayushjaink8 committed:
fixed backend test errors

ayushjaink8 commented:
@seanlip Please take a look. I have replied to all of your comments and incorporated all your suggesstions.

seanlip commented:
Hi @sagangwee -- had a question for you [here](https://github.com/oppia/oppia/pull/15460#discussion_r896609201), too. PTAL!

gp201 commented:
@ayushjaink8 I'm deassigning myself. Please assign me towards the end. This is done to prevent flakes from entering develop, Thanks!

ayushjaink8 committed:
merged commits and changed preferred_translation_language_code description

ayushjaink8 committed:
fixed lint check errors and front-end test errors

ayushjaink8 commented:
@seanlip I have accommodated all the requested changes. Kindly, have a look and approve this PR.

ayushjaink8 committed:
fixed trailing spaces lint check error

ayushjaink8 committed:
Merge branch 'oppia:develop' into alang

seanlip reviewed (approved):
Thanks @ayushjaink8! LGTM for codeowners.

ayushjaink8 commented:
@SAEb-ai @kevintab95 @srijanreddy98 @ashutoshc8101 @vojtechjelinek Kindly take a look and approve this PR.

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -247,6 +253,43 @@ export class ContributionOpportunitiesBackendApiService {
+  ): Promise<void> {
+    return this.userService.getUserInfoAsync().then(
+      (userInfo) => {
+        if (userInfo?.isLoggedIn()) {
```
Why can be `userInfo` `undefined`?

ayushjaink8 replied:
It can not be undefined. I have updated this and removed the `?`.

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -247,6 +253,43 @@ export class ContributionOpportunitiesBackendApiService {
+    return this.userService.getUserInfoAsync().then(
+      async(userInfo) => {
+        if (userInfo?.isLoggedIn()) {
+          const res = await this.http
+            .get<PreferredTranslationLanguageBackendDict>(
+              '/preferredtranslationlanguage').toPromise().catch(
+              () => emptyResponse);
```
```suggestion
          const res = (
            await this.http.get<PreferredTranslationLanguageBackendDict>(
              '/preferredtranslationlanguage'
            ).toPromise().catch(() => emptyResponse)
          );
```

ayushjaink8 replied:
I have updated this. Thanks.

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -247,6 +253,43 @@ export class ContributionOpportunitiesBackendApiService {
+              () => emptyResponse);
+          return res.preferred_translation_language_code;
+        } else {
+          return '';
```
Why we return empty string and not `null`?

ayushjaink8 replied:
If we use `null` then we need to add a lot of if-else statements to deal with the typescript errors. Also, I was trying to follow the general trend which was being followed in this file (like returning empty arrays `[]` instead of `null` in `fetchTranslatableTopicNamesAsync` and `fetchFeaturedTranslationLanguagesAsync`).

vojtechjelinek replied:
I think `''` is different from `[]`, we usually add stuff to lists but for strings we usually just set them and then work with them. We generally try not to use `''` except for some rare cases.

ayushjaink8 replied:
I have replaced the empty string with `null`.

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

ayushjaink8 committed:
incorporated requested suggestions

ayushjaink8 commented:
@vojtechjelinek I have replied to your comments. PTAL!

ayushjaink8 committed:
replaced empty string with null

ayushjaink8 committed:
Merge branch 'oppia:develop' into alang

ayushjaink8 commented:
@vojtechjelinek I have replaced the empty string with `null`. Kindly, have another look.

ayushjaink8 committed:
Merge branch 'oppia:develop' into alang

ayushjaink8 committed:
Merge branch 'oppia:develop' into alang

ayushjaink8 commented:
@vojtechjelinek PTAL

SAEb-ai commented on core/tests/protractor_desktop/contributorDashboard.js:
```diff
@@ -121,6 +121,25 @@ describe('Contributor dashboard page', function() {
       GERMAN_LANGUAGE);
   });
 
+  it('should persist user\'s preferred translation language across sessions',
+    async function() {
+      await users.login(USER_EMAILS[0]);
+      await contributorDashboardPage.get();
+      await contributorDashboardPage.navigateToTranslateTextTab();
+      await contributorDashboardTranslateTextTab.changeLanguage(
+        GERMAN_LANGUAGE);
+      await contributorDashboardTranslateTextTab.expectSelectedLanguageToBe(
+        GERMAN_LANGUAGE);
+      await users.logout();
+
+      await users.login(USER_EMAILS[0]);
+      await contributorDashboardPage.get();
+      await contributorDashboardPage.navigateToTranslateTextTab();
+      await contributorDashboardTranslateTextTab.expectSelectedLanguageToBe(
+        GERMAN_LANGUAGE);
+      await users.logout();
+    });
+
```
Can you please assure that the test is not flaky by following [this](https://github.com/oppia/oppia/wiki/End-to-End-Tests#checking-for-flakiness)?

ayushjaink8 replied:
@SAEb-ai I have added the screenshots of all the re-runs. 
Refer: https://github.com/oppia/oppia/pull/15460#issue-1233215184

SAEb-ai replied:
I think you have only pasted three screenshots for the successful runs of the suite `contributorDashboard`. But the commitment was for five. The first two screenshots are for `Oppiabot`. From the next time, please keep this in mind that you have to submit screenshots of the suite only in which you have added new tests.  

But the green signal in the commit messages of the first two screenshots ultimately shows that all the tests have passed. Hence, I will approve this. Thanks!!

ayushjaink8 committed:
Merge branch 'oppia:develop' into alang

vojtechjelinek commented on core/templates/pages/contributor-dashboard-page/services/contribution-opportunities-backend-api.service.ts:
```diff
@@ -247,6 +253,44 @@ export class ContributionOpportunitiesBackendApiService {
+    };
+    return this.userService.getUserInfoAsync().then(
+      async(userInfo) => {
+        if (userInfo?.isLoggedIn()) {
```
Why `?`?

ayushjaink8 replied:
Sorry, Left this one by mistake. I have already removed this `?` from the previous function. Will remove it from here as well.

vojtechjelinek reviewed (approved):
Thanks. Just a nit.

ayushjaink8 committed:
retrigger checks - 2

ayushjaink8 committed:
Merge branch 'alang' of https://github.com/ayushjaink8/oppia into alang

ayushjaink8 committed:
Merge branch 'oppia:develop' into alang

ayushjaink8 committed:
removed ?

ashutoshc8101 reviewed (changes requested):
Changes to the file `UpgradedSerivces.ts` are not required.

ayushjaink8 committed:
Merge branch 'oppia:develop' into alang

ayushjaink8 commented:
> Changes to the file `UpgradedSerivces.ts` are not required.

@ashutoshc8101. Actually, I am facing a typescript check error (shown below). Since we are adding `userService` in our component so we need to add that here as well.

Refer: https://github.com/oppia/oppia/runs/6865051223?check_suite_focus=true

Also, when we can reproduce this error locally:
![image](https://user-images.githubusercontent.com/75239620/174989063-74d60d88-acd2-40a3-ae34-9492c7077437.png)

ashutoshc8101 commented:
> > Changes to the file `UpgradedSerivces.ts` are not required.
> 
> @ashutoshc8101. Actually, I am facing a typescript check error (shown below). Since we are adding `userService` in our component so we need to add that here as well.
> 
> Refer: https://github.com/oppia/oppia/runs/6865051223?check_suite_focus=true
> 
> Also, when we can reproduce this error locally:
> ![image](https://user-images.githubusercontent.com/75239620/174989063-74d60d88-acd2-40a3-ae34-9492c7077437.png)
> 
> 

Understandable, its fine to keep these changes.

ashutoshc8101 reviewed (approved):
LGTM for codeowners

kevintab95 reviewed (approved):
LGTM for 2 codeowner files. Thanks @ayushjaink8!

gp201 reviewed (approved):
LGTM

ayushjaink8 commented:
@SAEb-ai If this PR looks good to you, then kindly merge this PR.

ayushjaink8 committed:
Merge branch 'oppia:develop' into alang

ayushjaink8 commented:
@srijanreddy98 PTAL and if this PR looks good to you then kindly merge this PR. :-)

seanlip commented:
@ayushjaink8 I think you have all the necessary approvals actually, so, merging this PR! Thanks for creating it :)

SAEb-ai commented:
@ayushjaink8 This PR has most probably caused a new flake in the end to end tests. The flake looks like this: 
```
- Expected $.length = 1 to equal 0.
Expected $[0] = Entry({ level: SEVERE, message: 'http://localhost:9001/build/webpack_bundles/vendors~lightweight_oppia_root~oppia_root.8451292e85351e5e5070.bundle.js 423:265 "ERROR" Error: Uncaught (in promise): TypeError: Failed to fetch
TypeError: Failed to fetch
    at b (http://localhost:9001/build/webpack_bundles/vendors~lightweight_oppia_root~oppia_root.8451292e85351e5e5070.bundle.js:3525:1919)
    at b (http://localhost:9001/build/webpack_bundles/vendors~lightweight_oppia_root~oppia_root.8451292e85351e5e5070.bundle.js:3525:1443)
    at http://localhost:9001/build/webpack_bundles/vendors~lightweight_oppia_root~oppia_root.8451292e85351e5e5070.bundle.js:3525:2754
    at e.invokeTask (http://localhost:9001/build/webpack_bundles/vendors~lightweight_oppia_root~oppia_root.8451292e85351e5e5070.bundle.js:3517:7859)
    at Object.onInvokeTask (http://localhost:9001/build/webpack_bundles/vendors~lightweight_oppia_root~oppia_root.8451292e85351e5e5070.bundle.js:1377:1589)
    at e.invokeTask (http://localhost:9001/build/webpack_bundles/vendors~lightweight_oppia_root~oppia_root.8451292e85351e5e5070.bundle.js:3517:7780)
    at t.runTask (http://localhost:9001/build/webpack_bundles/vendors~lightweight_oppia_root~oppia_root.8451292e85351e5e5070.bundle.js:3517:2934)
    at y (http://localhost:9001/build/webpack_bundles/vendors~lightweight_oppia_root~oppia_root.8451292e85351e5e5070.bundle.js:3517:10199)', timestamp: 1656273399306, type: '' }) to equal undefined.
```
Few examples of the test runs that have this flake(develop branch):

- https://github.com/oppia/oppia/runs/7062909767?check_suite_focus=true#step:10:3248
- https://github.com/oppia/oppia/runs/7062909767?check_suite_focus=true#step:10:3248
- https://github.com/oppia/oppia/runs/7087746548?check_suite_focus=true#step:13:297

The Automated QA Team went through the commits on the develop branch to find out the suspicious PR that first introduced the flake in the end to end tests and we found this. Will you be able to provide a quick fix for this? If not, then we will be opening a revert PR for this soon. Thanks!!

ayushjaink8 commented:
Hi @SAEb-ai!

I have checked a few CI checks and found out that my PR wasn't the first one who encountered this flake. For instance,
- https://github.com/oppia/oppia/runs/7058214350?check_suite_focus=true

This error persisted before this PR was merged. Moreover, I am pretty sure that I haven't touched creatorDashboard in any sense via my PR. Also, I have tried to look into the details and found out that the same flake is reoccurring in many similar forms like:
- https://github.com/oppia/oppia/runs/7155711174?check_suite_focus=true
- https://github.com/oppia/oppia/runs/7203248155?check_suite_focus=true

Apart from this, I am having an intuition that the core of the problem has something to do with this PR error (develop branch):
https://github.com/oppia/oppia/runs/6430788730?check_suite_focus=true
PR: https://github.com/oppia/oppia/pull/15450
